### ISSUE TITLE
Automaticaly update

### ISSUE THREAD

**Issue by GonzaloSP** [odeke-em/drive#446]
Hi,

I was wondering if someone has developed a bash script to update every x minutes a google drive directory?

Thanks

P.S. The script works perfectly

**Comment by canpolat** [comment]
I did something similar using **cron**. Are you looking for a script or do you want to share your script?

**Comment by GonzaloSP** [comment]
I did write a script but it is very simple. I am trying to imitate the behavior of Google Drive for Windows.

So, I did:

#!/bin/bash
cd /shareDirectory
drive push -fix-clashes --no-prompt
drive pull -fix-clashes --no-prompt

And I added to the cron of the root user . If someone has a better script and wanna share please post it.

Thanks

**Comment by canpolat** [comment]
Mine is even simpler than yours, so it wouldn't help. But if you want to automatic synchronization, I would suggest looking into `inotify`. Google gave me this: [How to use inotify-tools to trigger scripts on filesystem events](http://techarena51.com/index.php/inotify-tools-example/). In this example, `rsync` is used to synchronize modify files, but the idea is the same. Probably it's a good idea to do a full synchronization when the PC is booted first and then go into a infinite wait loop for notifications. Unfortunately I cannot comment on the performance and scalability of such a solution. But if you manage to trigger a `push` per file (and not for the whole directory) I believe it may be less painful. If you happen to solve the problem please share it with us.

**Comment by odeke-em** [comment]
Hello again @GonzaloSP  and @canpolat.

Awesome that y'all customizing it up. So as you'd guess am staying away from auto-sync for a bit until I have enough time to experiment properly with it :) @canpolat nice -- if you'd like a Go solution, there is a Go library github.com/go-fsnotify/fsnotify. I actually used fsnotify to build a simple watch and notify utility for a friend https://github.com/odeke-em/watnot, you could adapt it for that. Also ages ago, I built some cross platform directory watching utility but unfortunately it is closed source :( but probably over the christmas break I might make an OSS equivalent :)
Thanks for sharing with everyone your solutions, this is great to see.

**Comment by bellera** [comment]
Based on:

http://wolf911.us/wgo/info/doku.php?id=install_grive_on_sabayon#script
http://elbauldelprogramador.com/sincronizar-google-drive-en-linux-en-4-pasos-actualizacion/

I'm wondering if something like:

while true
   do
       inotifywait -e modify -e move -e create -e delete -r  /home/my_user/gdrive
       cd /home/my_user/gdrive
       drive push -fix-clashes -quit
       drive pull -fix-clashes -quit
  done

will be a correct choice.

For pulling it's necessary another process

**Comment by canpolat** [comment]
@bellera, `push` and `pull` commands should have the `-quiet` flag, not `-quit`. Otherwise it looks fine to me. But, I don't know how it would scale. Has anyone tried this?

**Comment by bellera** [comment]
Oops, sorry for the miss spelling!

I didn't tried yet. I just finishing the migration of my users from local server to Google Drive, helped by "drive".

Now I have to choose a client "transparent" solution for Ubuntu Desktop. A script like this could be a solution. Or a GUI solution. However, GUI solutions for Linux are not free at the moment.

**Comment by canpolat** [comment]
BTW: One problem I see here is that when a user is working on a document, `inotifywait` will trigger every time the document is saved. So, the script will go into a `push`-`pull` circle on every save (which is not ideal). If the user is anything like me (I save a document every few seconds) and the folder contains too many files, bad things may happen.

I believe I would go for a solution where the user manually triggers synchronization (maybe in addition to a cron job that runs after midnight). But I don't know your requirements.

**Comment by GonzaloSP** [comment]
canpolat, why do you think it wouldn't be ideal? (taking aside the bandwidth consumption and I/O requests)
It is what the google drive app for windows does... and I am happy with the way it works. 
In fact, a great deal of my day I spend it on writing, and I my FreeOffice-Word app saves the file every 3 minutes. In that way I will be updating my files every 3 minutes, I don't see how that could generate a problem....?

Thanks

**Comment by canpolat** [comment]
@GonzaloSP it depends on the size of the directory and frequency of saves, I suppose. I have directories that take minutes to resolve the differences. When I work on documents, I save every 10 seconds. In my case, the script will not trigger every 10 seconds (because it would be in resolving phase after the first trigger) but it will never stop either. It will be in a never ending resolving loop.

I guess, what I'm trying to say is, resolving the differences in the whole folder is not optimum. I would recommend finding the actual modified file and `push`ing that single file.

**Comment by cgogolin** [comment]
I am not sure I fully understand the behavior of pull and push, but from by experience I see the following rather severe problem with the script of @bellera : Suppose a file f1 has been modified or added to the remote Google Drive and then a user edits or adds some other local file f2. The script will trigger and the `drive push -fix-clashes -quiet` will overwrite or delete the new/modified remote file f1 with the local version before the call of `drive pull -fix-clashes -quiet` could download the more up to date remote version. If the two are switched around, then f1 is overwritten instead by the old remote version. Am I missing something? Can this bis done? Is there some way to push/pull only "newer" files?

**Comment by intrd** [comment]
Thank you guys,
Following your tips i was able to make a script for auto-sync (push and pull), its working very well in a directory with few files..

`#!/bin/bash
while true
do
	cd /home/intrd/cloud/cloud

	if inotifywait -t2 -e modify -e move -e create -e delete -r /home/intrd/cloud/cloud; then
		cd /home/intrd/cloud/cloud
		drive push -fix-clashes -quiet
		drive pull -fix-clashes -quiet
		echo "PUSHED"
	fi

	drive pull -fix-clashes -quiet
	drive push -fix-clashes -quiet
	echo "PULLED"
done`

**Comment by odeke-em** [comment]
So Google actually has provided a Real time API https://developers.google.com/google-apps/realtime/overview and it works with drive/v2 and drive/v3 https://developers.google.com/drive/v2/reference/realtime. It keeps a persistent
connection to the cloud and on any changes to particular files, those get pushed down
to the listener and then they can decide on what to do with it. Also before that, they
provided the changes API https://developers.google.com/drive/v2/reference/changes,
which I had started integrating but my needs diverged a bit so I haven't tinkered much more with
it https://github.com/odeke-em/drive/blob/master/src/remote.go#L113-L142

So basically if someone would like to implement a real time watching sync, it is quite possible.
I personally haven't had the time to work on it for OSS, nor have I had motivation for it, as am
swamped with a full time job, other projects, and soon school coming back up.
However, if things can work out, IMO it would be a great project!

**Comment by gniding** [comment]
> Is there some way to push/pull only "newer" files?

There is at least one way. See https://github.com/odeke-em/drive#pushing

Excluding certain operations can be done both for pull and push by passing in flag -exclude-ops
e.g drive pull -exclude-ops "delete,update" vines

**Comment by ErwinP** [comment]
Hi,

I'm experiencing something weird here. I made a script to sync and backup specific folders and that script works great when I execute it myself.
However, if I add it as a cronjob to either my cron or the roots cron, it fails to synchronize. I can see from my "log" that it executes the command, but than immediatly continues to the next, as if the drive command gave some error. I however don't know the nature of this error, as the scripts works just fine when I execute it myself.

Any idea's on how to troubleshoot?

Thank you in advance!
Erwin

**Comment by davidhcefx** [comment]
As @cgogolin have mentioned, there is no `fetch` command in **drive** as in **git**. Therefore, if *file1* has been updated remotely and *file2* has been updated locally, we couldn't blindly run `drive pull` nor `drive push`, because either one would erase one of the updated file.

Since there is a method of using `inotifywait` to find out which file has been modified *locally*, we could first push that specific file `drive push $TheFileName`, and then `drive pull` to merge. We have to merge them this way, because we don't know which file has been modified *remotely*. (Unless we use the Real time API Google provided, which was described in @odeke-em 's comment)

However, the problem gets more complicated when the SAME *remote* and *local* files updated. Which version shall we keep? Therefore, the above strategy would fail in this situation, since we couldn't get the notification of remote file changes.

Let's assume that we also utilized the Real-time Google API in order to find out which file has been updated remotely. When the file updated *remotely*, we `drive pull $TheFileName` in the first place. After that, let's say it updated *locally*, so we then run `drive push $TheFileName`. But what would happen if someone are modifying a file *remotely* while we are also modifying it *locally* AT THE SAME TIME? Let's say the *remote side* pressed `Ctrl+S` first, would our *local* file be overwritten?(certainly yes) What would our unsaved text go? Although we have text-editing buffer to still keep the data, we might not write it out directly. (I use `nano a.txt` and `echo 123 > a.txt` for an example)

It seems quite challenging if we want **auto-sync** to be able of updating on both sides. However, if we simply use Google Drive as a backup container and we **only updates on the local side**, it's relatively easy to implement!

Here's a bash script that uses `inotifywait` to find out which file to update. Please run `sudo apt install inotify-tools` first.

	#! /bin/bash
	# drive-daemon
	# Script for `drive push` automatically upon file update
	# Written by davidhcefx, 13,Feb,2018

	watch0=
	file0=
	lastmodtime=0
	lastmodfile=/null
	exec 1>>~/drived.log   # Change to where you want to store log file
	exec 2>&1
	cd ~/your/drive/folder  # Change to your own path

	inotifywait -m -e create -e delete -e move -e modify -e attrib \
		--exclude '.swp$' \  # Add more ignore rules if you want
		-r ./ | while read watch event file; do
		echo "$(date +%F\ %T): $watch $event $file"
		case $event in
		    CREATE*)
				drive push -no-prompt $watch$file
		    ;;
		    DELETE*)
				drive trash -quiet $watch$file
		    ;;
		    MOVED_FROM*)
				watch0=$watch
				file0=$file
		    ;;
		    MOVED_TO*)
				if [ $watch0 != $watch ]; then # move
					drive move $watch0$file0 $watch
				fi
				if [ $file0 != $file ]; then # rename
					drive rename -local=0 $watch$file0 $watch$file
				fi
		    ;;
		    MODIFY)
				cur=`date +%s`
				if (( lastmodtime <= cur-3 )); then  # 3 sec max interval
					drive push -no-prompt $watch$file
					lastmodtime=$cur
					lastmodfile=$watch$file
				elif [ "$lastmodfile" != "$watch$file" ]; then
					drive push -no-prompt $lastmodfile $watch$file
					lastmodtime=$cur
					lastmodfile=$watch$file
				fi
		    ;;
		    ATTRIB) # i.e. chmod
				drive push -no-prompt $watch$file
		    ;;
		esac
	done
